assembl/static2/js/app/client.js

```diff
@@ -40,16 +46,28 @@ const networkInterface = createNetworkInterface({
 networkInterface.use([
   {
     applyMiddleware: function (req, next) {
-      Sentry.addBreadcrumb({
-        category: 'graphql',
-        message: `GraphQL operation: ${req.request.operationName}`,
-        data: {
-          variables: req.request.variables
-        },
-        level: 'info'
-      });
+      if (!req.options.headers) {
+        req.options.headers = {}; // Create the header object if needed.
+      }
+      co(async () => {
+        if (useCSRFProtection === 'true') {
+          const responseToken = await getCSRFToken();
```

Comment:
Okay, this will make `bond` ms a bottleneck, because every graphQL call is a POST, and you're fetching a new token for every query and mutation. You should peek internally into the request to see if it's a mutation or a query, and only `getCSRFToken()` when it's a mutation.

Reply:
I agree but I'm not sure if that covers what we're trying to achieve. For example the OWASP standards on CSRF don't care if it's graphql and a query or mutation, it cares only that POST requests are not susceptible. Also, we don't have a choice, in order to add CSRF protection we will have a slow down somewhere for the extra round-trips. Happy to make a change but I would suggest we release this as a first version and see what the penalty looks like.

Reply:
Hmm, yeah OWASP is covering the base HTTP protection level. Okay, let's pay the price for now and see if we can win the time back elsewhere. We'll keep a close watch on response times. We're already quite terrible (+2s). There are many other places to optimize the response speed (code-splitting, anyone?)